### PR TITLE
Implement de-duplication of NuGet and VSIX analyzer execution and diagnostics

### PR DESCRIPTION
**Pull request by mavasani**
Implements the design in #18818 for analyzers. De-duping for code actions will be implemented in a follow-up PR.

1. We now execute only those VSIX analyzers which report at least one unique diagnostic ID that is not reported by any NuGet based project analyzers.
2. Additionally, we filter out the diagnostics of IDs reported by VSIX analyzers, which also have a corresponding NuGet analyzer reporting the same diagnostic ID.

First commit implements the unit tests as per the design mentioned in #18818 - all the added unit tests failed after this commit
Second commit implements the above logic in the IDE diagnostic analyzer engine.

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs [hunk 1]
```diff
@@ -85,6 +85,11 @@ internal sealed partial class DiagnosticAnalyzerInfoCache
         /// </remarks>
         private readonly ConditionalWeakTable<DiagnosticAnalyzer, DiagnosticDescriptorsInfo> _descriptorsInfo;
 
+        /// <summary>
+        /// Information about host analyzers that can be skipped for the given project ID.
+        /// </summary>
+        private readonly ConditionalWeakTable<ProjectId, ISkippedAnalyzersInfo> _skippedAnalyzersInfo;
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs]: ❔ Any reason to not use the same `IReadOnlyList<AnalyzerReference>` key that `CodeFixService` uses?

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs [hunk 2]
```diff
@@ -145,12 +146,23 @@ private static string GetAssemblyQualifiedName(Type type)
...
+                {
+                    continue;
+                }
+
                 var result = new DiagnosticAnalysisResultBuilder(project, version);
+                var diagnosticIdsToFilter = ImmutableArray<string>.Empty;
+                if (skippedAnalyzersInfo.FilteredDiagnosticIdsForAnalyzers.TryGetValue(analyzer, out var ids))
+                {
+                    diagnosticIdsToFilter = ids;
+                }
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Diagnostics/Extensions.cs]:
```suggestion
                var diagnosticIdsToFilter = skippedAnalyzersInfo.FilteredDiagnosticIdsForAnalyzers.GetValueOrDefault(
                    analyzer,
                    ImmutableArray<string>.Empty);
```

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs [hunk 3]
```diff
@@ -177,6 +191,30 @@ private static string GetAssemblyQualifiedName(Type type)
...
+            {
+                if (diagnosticIdsToFilter.Contains(diagnostic.Id))
+                {
+                    continue;
+                }
+
+                builder.Add(diagnostic);
+            }
+
+            return builder.ToImmutableAndFree();
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Diagnostics/Extensions.cs]: Any reason to not go with the easy approach here?

```suggestion
            return diagnostics.RemoveAll(diagnostic => diagnosticIdsToFilter.Contains(diagnostic.Id));
```

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs [hunk 4]
```diff
@@ -86,8 +86,11 @@ public InProcOrRemoteHostAnalyzerRunner(IAsynchronousOperationListener operation
                 var asyncToken = _asyncOperationListener.BeginAsyncOperation(nameof(AnalyzeInProcAsync));
                 var _ = FireAndForgetReportAnalyzerPerformanceAsync(project, client, analysisResult, cancellationToken).CompletesAsyncOperation(asyncToken);
 
+                // get skipped analyzers info
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs]: Comment is redundant

```suggestion
```

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs]: We only have the project ID in the project removal callback when we clear this cache entry.

File: src/Workspaces/Core/Portable/Diagnostics/Extensions.cs [hunk 5]
```diff
@@ -145,12 +146,23 @@ private static string GetAssemblyQualifiedName(Type type)
...
+                {
+                    diagnosticIdsToFilter = ids;
+                }
 
                 foreach (var (tree, diagnosticsByAnalyzerMap) in analysisResult.SyntaxDiagnostics)
                 {
                     if (diagnosticsByAnalyzerMap.TryGetValue(analyzer, out diagnostics))
                     {
+                        diagnostics = diagnostics.Filter(diagnosticIdsToFilter);
                         Debug.Assert(diagnostics.Length == CompilationWithAnalyzers.GetEffectiveDiagnostics(diagnostics, compilation).Count());
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/Core/Portable/Diagnostics/Extensions.cs]: It should be after as the Filter API removes the diagnostics from partially skipped analyzers - i.e. ones which have a subset of diagnostic IDs to be skipped.

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs]: Wouldn't it clear itself if we use CWT?

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs [hunk 6]
```diff
@@ -43,9 +44,15 @@ public DiagnosticDescriptorsInfo(ImmutableArray<DiagnosticDescriptor> supportedD
             }
         }
 
+        /// <summary>
+        /// Information about host analyzers that can be skipped for the given project analyzers.
+        /// </summary>
+        private readonly ConditionalWeakTable<IReadOnlyList<AnalyzerReference>, ISkippedAnalyzersInfo> _skippedAnalyzersInfo;
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs]: @sharwell CWT feedback addressed.

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs [hunk 7]
```diff
@@ -114,5 +121,29 @@ public bool IsAnalyzerSuppressed(DiagnosticAnalyzer analyzer, Project project)
...
+            {
+                return skippedAnalyzersInfo;
+            }
+
+            var createValueCallback = new ConditionalWeakTable<IReadOnlyList<AnalyzerReference>, ISkippedAnalyzersInfo>.CreateValueCallback(
+                _ => SkippedHostAnalyzersInfo.Create(project, hostAnalyzers, this));
+            return _skippedAnalyzersInfo.GetValue(project.AnalyzerReferences, createValueCallback);
+        }
+
+        public ISkippedAnalyzersInfo GetOrCreateSkippedAnalyzersInfo(Project project, IEnumerable<AnalyzerReference> hostAnalyzers)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerInfoCache.cs]: This API is used for computing this info in OOP, as we don't have `HostDiagnosticAnalyzers` available in OOP.

**Comment by mavasani** [comment]
Thank you @sharwell

**Pull request merged**